internal/servers/destinations.go

```diff
@@ -86,36 +94,42 @@ func (s *DestinationServer) Write2(msg pb.Destination_Write2Server) error {
 
 	for {
 		r, err := msg.Recv()
-		if err != nil {
-			if err == io.EOF {
-				close(resources)
-				if err := eg.Wait(); err != nil {
-					return fmt.Errorf("got EOF. failed to wait for plugin: %w", err)
-				}
-				return msg.SendAndClose(&pb.Write2_Response{})
+		if err == io.EOF {
+			close(resources)
+			if err := eg.Wait(); err != nil {
+				return addErrorDetails(errors.New("plugin write failed"), err.Error())
 			}
+			return msg.SendAndClose(&pb.Write2_Response{})
+		}
+		if err != nil {
 			close(resources)
 			if err := eg.Wait(); err != nil {
-				s.Logger.Error().Err(err).Msg("got error. failed to wait for plugin")
+				return addErrorDetails(errors.New("plugin write failed"), err.Error())
 			}
 			return fmt.Errorf("failed to receive msg: %w", err)
 		}
 		var resource schema.DestinationResource
-		if err := json.Unmarshal(r.Resource, &resource); err != nil {
+		if unmarshalError := json.Unmarshal(r.Resource, &resource); unmarshalError != nil {
 			close(resources)
-			if err := eg.Wait(); err != nil {
-				s.Logger.Error().Err(err).Msg("failed to unmarshal resource. failed to wait for plugin")
+			errWithDetails := addErrorDetails(errors.New("failed to unmarshal resource"), unmarshalError.Error())
+			if writeError := eg.Wait(); writeError != nil {
+				errWithDetails = addErrorDetails(errors.New("plugin write failed and failed to unmarshal resource"), writeError.Error())
+				errWithDetails = addErrorDetails(errWithDetails, unmarshalError.Error())
 			}
-			return status.Errorf(codes.InvalidArgument, "failed to unmarshal resource: %v", err)
+			return errWithDetails
 		}
 		select {
 		case resources <- resource:
 		case <-ctx.Done():
 			close(resources)
 			if err := eg.Wait(); err != nil {
-				s.Logger.Error().Err(err).Msg("failed to wait")
+				return addErrorDetails(errors.New("plugin write failed"), err.Error())
+			}
+			ctxError := ctx.Err()
+			if ctxError == nil {
+				return nil
 			}
-			return ctx.Err()
+			return addErrorDetails(errors.New("context error"), ctxError.Error())
```

Comment:
This is only an example, we should be using https://github.com/cloudquery/plugin-sdk/pull/486 instead